### ISSUE TITLE
ty does not match variadic argument to variadic parameter

### ISSUE THREAD

**Issue by dhruvmanila** [astral-sh/ty#1236]
### Summary

In the following example:
```py
def f(x: int, *args: str) -> int:
    return 1


def _(x: list[int]):
    reveal_type(f(*x))
```

The above code should raise a type error stating that `int` cannot be assigned to `str` for the variadic parameter but ty doesn't do that.

This is because ty doesn't try to match variadic argument to the variadic parameter in certain scenarios like the above. This is due to missing logic in `ArgumentMatcher::match_variadic`.
